config/routes.rb

```diff
@@ -1,3 +1,13 @@
+# @note I'm not sure where best to declare this method, which is only used here
+def die_if_unsupported_configuration!
+  return if ENV.fetch('SERVER_SOFTWARE', '').match(/Apache.*Phusion_Passenger/).nil?
+  helpful_error_message = 'ERROR: cannot deploy with realtime notifications atop Passenger + Apache. ' \
+                          'Disable this feature in config/initializers/hyrax.rb by setting config.realtime_notifications to false'
+  # It's easy to miss this error message in the server log, so also put it in the Rails log
+  Rails.logger.error(helpful_error_message)
+  raise helpful_error_message
+end
+
```

Comment:
please don’t hard-code a sniff for apache-passenger disabling a feature, what if a future passenger supports it?  At least make it possible to override with config, instead of hard-coding in the assumption that passenger will not support this for all time, meaning if it ever does, an upgrade to hyrax will be required to allow it to.

Reply:
It's not disabling a feature. It's saying that if you proceed with this feature enabled, it will lock up passenger and crash the app.

During the meeting today, the consensus was...
  1. keep notifications using ActionCable turned on by default.
  1. Sniff out the 1 known config setting that is incompatible, and if found, require that the option be turned off.

This is due to the fact that we couldn't find a rescuable exception after attempting to mount the ActionCable server. Without a rescuable exception, we have to stop passenger from freezing before hand.

Reply:
@jrochkind @mjgiarlo What about something like this?...

```ruby
if Hyrax.config.realtime_notifications?
  using_passenger = ENV.fetch('SERVER_SOFTWARE', '').match(/Apache.*Phusion_Passenger/).nil?
  use_passenger_anyway = ENV.fetch('USE_NOTIFICATIONS_WITH_APACHE_PASSENGER', false)
  raise ApachePassengerIncompatibleWithNotifications if (using_passenger && !use_passenger_anyway)

  namespace :notifications do
    # WebSocket for notifications
    mount ActionCable.server => 'endpoint', as: :endpoint
  end
end

# And then in lib/hyrax/errors.rb
module Hyrax
# ...
  class ApachePassengerIncompatibleWithNotifications < Hyrax::Error
    def initialize
      msg = 'ERROR: cannot deploy with realtime notifications atop Passenger + Apache. ' \
                  + 'Disable this feature in config/initializers/hyrax.rb by setting config.realtime_notifications to false'
      super(msg)
    end
  end
# ...
end
```

And then we wouldn't need the `die_if_unsupported_configuration` method.

Reply:
@afred could you take a look at the current solution, which is mostly in line with what you're suggesting? The current solution is to coerce the value of this config to false if we sniff Passenger + Apache, and then warn the user that they should set this to false to stop seeing the warning. If the user decides that want to proceed anyway, they can override `Hyrax.config.realtime_notifications?` in `config/initializers/hyrax.rb`. Cool?

Reply:
totally cool